src/librustc/traits/error_reporting.rs

```diff
@@ -767,32 +831,87 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         err.emit();
     }
 
-    fn report_arg_count_mismatch(&self,
-                                 span: Span,
-                                 found_span: Option<Span>,
-                                 expected: usize,
-                                 found: usize,
-                                 is_closure: bool)
-        -> DiagnosticBuilder<'tcx>
-    {
+    fn report_arg_count_mismatch(
+        &self,
+        span: Span,
+        found_span: Option<Span>,
+        expected: usize,
+        expected_tuple: Option<usize>,
+        found: usize,
+        closure_args: Option<(Vec<String>, Vec<Option<String>>)>,
+        is_closure: bool
+    ) -> DiagnosticBuilder<'tcx> {
+        use std::borrow::Cow;
+
+        let kind = if is_closure { "closure" } else { "function" };
+
+        let args_str = |n| format!(
+                "{} argument{}",
+                n,
+                if n == 1 { "" } else { "s" }
+            );
+
         let mut err = struct_span_err!(self.tcx.sess, span, E0593,
-            "{} takes {} argument{} but {} argument{} {} required",
-            if is_closure { "closure" } else { "function" },
-            found,
-            if found == 1 { "" } else { "s" },
-            expected,
-            if expected == 1 { "" } else { "s" },
+            "{} takes {}, but {} {} required",
+            kind,
+            if expected_tuple.is_some() {
+                Cow::from("multiple arguments")
+            } else {
+                Cow::from(args_str(found))
+            },
+            if expected_tuple.is_some() {
+                Cow::from("a tuple argument")
```

Comment:
I think I figured out what is bothering me about this wording. The construction is not quite parallel. It says "multiple arguments" are expected, but then says "but a tuple argument", which is not obviously the opposite. Also, our usual construction is "expected X, but found Y", whereas the message here says "found X, but Y is required" (pre-existing, I know, but I still find it a bit surprising).

Perhaps this?

**expected a closure that takes a single tuple as argument but found a closure that takes N distinct arguments**

Maybe that's too wordy. We could tweak it, though it goes a *bit* away from the "expected X but found Y" form:

**expected closure to take a single tuple as argument but it takes N distinct arguments**

or, better:

**closure is expected to take a single tuple as argument, but it takes N distinct arguments**

(I would probably change the existing message that doesn't mention tuples to match, as well.)

Reply:
It looks better, thanks! Reworded the message to the last one.